improve: sync mpackage modules

Edru2 opened the pull request:
#### Brief overview of PR changes/additions
Mpackage/Zip modules can be synced now

#### Motivation for adding to Mudlet
It wasn't possible to sync mpackage/zip modules

#### Other info (issues closed, discussion etc)

#### Release post highlight
Mpackage/Zip modules can also be synced now

Edru2 committed:
allow to save modules if not in syncmode

Edru2 committed:
allow syncing of mpackage modules

allows syncing of mpackage modules but has still some issues and leads to crash sometimes (need to investigate)

Edru2 committed:
try to solve some clang-tidy issues

Edru2 committed:
different approach only load xml

as it only loads the xml if a package gets synced it doesn't need to read the zip file at all
(not finished as it's still a bit repetitive in the code)

Edru2 committed:
Merge branch 'development' into save-modules-if-not-in-sync-mode

Edru2 committed:
delete xml in zip file before adding new one

overwriting xml file directly didn't work for some reason (also caused empty xml files)

Edru2 committed:
revert saving if not in sync mode

vadi2 commented:
Does https://github.com/Mudlet/Mudlet/blob/development/src/Host.cpp#L1867-L1871 need an update?

Edru2 commented:
Unfortunately it is not as ready as I thought yet. Found a strange issue that it is sometimes only syncing in one direction. Converting it to a draft again.

Edru2 committed:
potential fix for issues (waiting for profile save finish)

demonnic commented:
If you decide to revive this PR, you'll need to merge development back into it to pick up the danger checks

Edru2 committed:
Merge branch 'development' into sync-mpackage-modules

Edru2 committed:
suggestions

Edru2 committed:
simplification

Edru2 committed:
handle issues with asynchronous writing of xml

Edru2 committed:
space

Edru2 committed:
suggestions

demonnic commented:
so if I understand correctly when using this version I should be able to register a .mpackage module, update it, and have it save back to the original file? All the files in the moduleBackup folder seem to be the same, and the original file also seems untouched. Restarting Mudlet also seems to load up the original file.

Edru2 commented:
🤔🤔 Did you enable module sync? It won't do anything if sync is not enabled

22 Oct 2021 22:11:17 Damian Monogue ***@***.***>:

> so if I understand correctly when using this version I should be able to register a .mpackage module, update it, and have it save back to the original file? All the files in the moduleBackup folder seem to be the same, and the original file also seems untouched. Restarting Mudlet also seems to load up the original file.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub[https://github.com/Mudlet/Mudlet/pull/5475#issuecomment-949930241], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AON67DHLJOTQYHGX66A462TUIHAOBANCNFSM5FGEXI5A].
> Triage notifications on the go with GitHub Mobile for iOS[https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675] or Android[https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub].  [data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEgAAABICAYAAABV7bNHAAAAAXNSR0IArs4c6QAAAARzQklUCAgICHwIZIgAAAArSURBVHic7cEBDQAAAMKg909tDjegAAAAAAAAAAAAAAAAAAAAAAAAAAA+DFFIAAEctgHwAAAAAElFTkSuQmCC###24x24:true###][Tracking image][https://github.com/notifications/beacon/AON67DGRGVWKXEPI4SQJNIDUIHAOBA5CNFSM5FGEXI5KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOHCPMSAI.gif]
>

demonnic commented:
I had sync checked in the module manager, and it was syncing the changes between profiles, it just didn't persist through restart.

Edru2 commented:
Warning for anyone testing this.
Be sure to disable sync on your mpackage/zip modules before going back to an older/other version of Mudlet because it will possibly empty the zip/mpackage modules if the profile is saved in an other version if sync is enabled.

demonnic commented:
I have retested now, and I'm not sure what happened the first time, if I had another Mudlet open or what, but I've retried 10 times and 10/10 times it has worked. 

I did run afoul of what Edru just warned of, and had to remuddle the pkg I was testing with because it got zilched out.

vadi2 commented:
Can we fix this? People don't have to be testers - they could want to downgrade to an earlier Mudlet version and losing their modules wouldn't be ideal.

vadi2 commented on src/Host.cpp:
```diff
@@ -519,29 +520,33 @@ void Host::saveModules(int sync, bool backup)
         if (backup) {
             QString time = QDateTime::currentDateTime().toString("yyyy-MM-dd#HH-mm-ss");
-            savePathDir.rename(filename_xml, savePath + moduleName + time); //move the old file, use the key (module name) as the file
+            QFile::copy(filename_xml, savePath + moduleName + time);
```
Renaming a file on disk is typically much faster than creating a copy, this is a potential Mudlet freeze for people with lots of modules. Need to test. It would be better if we made a list of all of the files to copy and then copied them in a separate thread (like the zipping/etc is done asynchronously).

Edru2 replied:
> Renaming a file on disk is typically much faster than creating a copy

This is true but not in this case (mpackage/zip) as recreating the whole archive is definitely slower then just copying the previous one to the backup folder and then only changing the archives xml file.

btw. the last commit should make it async anyway

vadi2 commented on src/Host.cpp:
```diff
@@ -519,29 +520,33 @@ void Host::saveModules(int sync, bool backup)
+            QObject *obj = new QObject(this);
+            connect(this, &Host::profileSaveFinished, obj, [=]() {
+                slot_updateModuleZips(entry, moduleName);
+                obj->deleteLater();
```
Good idea, but I wonder because `obj` will be removed at the end of the method, will the slot still _always_ run? I'll double-check.

Edru2 replied:
I tested it and it won't run after being deleted. See also documentation https://doc.qt.io/qt-5/qobject.html#disconnect
I don't use this in this PR anymore

vadi2 commented on src/Host.cpp:
```diff
@@ -563,46 +568,52 @@ void Host::slot_reloadModules()
-        QStringList entry = it.value();
-        QString moduleName = it.key();
-        QString filename_xml = entry[0];
+    QString zipName = entry[0];
```
We only seem to pass the `entry` for the `zipName` - let's just pass the zipName then instead? It'll be a lot more obvious what is being passed in.

vadi2 commented on src/Host.cpp:
```diff
@@ -563,46 +568,52 @@ void Host::slot_reloadModules()
-        QString filename_xml = entry[0];
+    QString zipName = entry[0];
+    zip* zipFile = nullptr;
+    if (zipName.endsWith(QStringLiteral("mpackage"), Qt::CaseInsensitive) || zipName.endsWith(QStringLiteral("zip"), Qt::CaseInsensitive)) {
```
If you invert this `if`, you can avoid a whole level of nesting. Nicer to read on side by side diffs especially:

![image](https://user-images.githubusercontent.com/110988/138583629-14f6d553-a1eb-408e-b6b9-7a992a4feacc.png)

vadi2 commented on src/Host.cpp:
```diff
@@ -563,46 +568,52 @@ void Host::slot_reloadModules()
+        zip_delete(zipFile, xmlIndex);
+        struct zip_source* s = zip_source_file(zipFile, filename_xml.toUtf8().constData(), 0, -1);
+        if (mudlet::debugMode && s == nullptr) {
+            TDebug(QColor(Qt::white), QColor(Qt::red)) << tr("Failed to open xml file \"%1\" to save to module %2. Error message was: \"%3\".",
```
```suggestion
            TDebug(QColor(Qt::white), QColor(Qt::red)) << tr("Failed to open xml file \"%1\" inside module %2 to update it. Error message was: \"%3\".",
```

Is this better?

vadi2 commented on src/Host.cpp:
```diff
@@ -563,46 +568,52 @@ void Host::slot_reloadModules()
+        if (mudlet::debugMode && s == nullptr) {
+            TDebug(QColor(Qt::white), QColor(Qt::red)) << tr("Failed to open xml file \"%1\" to save to module %2. Error message was: \"%3\".",
+                                                             // Intentional comment to separate arguments
+                                                             "This error message will appear when the xml file cannot be added to the module for some reason.")
```
```suggestion
                                                             "This error message will appear when the xml file inside the module zip cannot be updated some reason.")
```

vadi2 commented on src/Host.cpp:
```diff
@@ -632,6 +643,40 @@ void Host::reloadModule(const QString& reloadModuleName)
+            }
+        }
+    }
+    //iterate through mInstalledModules again and reset the entry flag to be correct.
```
What do you mean by the entry flag?

Edru2 replied:
Honestly I'm not sure, I just took it from the previous reloadModule function.
I rewrote the function to directly use the reloadModule function now

vadi2 commented on src/Host.cpp:
```diff
@@ -640,12 +685,6 @@ std::pair<bool, QString> Host::changeModuleSync(const QString& moduleName, const
 
     if (mInstalledModules.contains(moduleName)) {
         QStringList moduleStringList = mInstalledModules[moduleName];
-        QFileInfo moduleFile = moduleStringList[0];
```
Is this the right PR to remove this in? Maybe make another?

Edru2 replied:
hmm what do you mean? if I don't remove this mpackage/zip files won't be syncable in the modulemanager and the purpose of this PR is to make them syncable

vadi2 replied:
OK, nevermind me then

vadi2 commented on src/Host.cpp:
```diff
@@ -1730,7 +1774,7 @@ bool Host::installPackage(const QString& fileName, int module)
             } else {
                 mInstalledPackages.append(packageName);
             }
-            reader.importPackage(&file2, packageName, module); // TODO: Missing false return value handler
```
Better to keep the TODO. Sorry it's hitting your PR, but could you put this back and file an issue for this?

Edru2 committed:
Async module save/sync

Modules(xml and zip/mpackage) can now be saved and synced in an async way which should prevent hanging of the application

Edru2 committed:
previous suggestions

put the todo back and apply the suggestions

Edru2 committed:
clang tidy suggestions

Edru2 commented:
:thinking:  what happened? It compiles without issues on my machine.

Edru2 committed:
Merge branch 'development' into sync-mpackage-modules

Edru2 committed:
Merge branch 'development' into sync-mpackage-modules

Edru2 committed:
Update Host.cpp

vadi2 commented:
Got it sorted?

Edru2 commented:
> Got it sorted?

yes, I just had to merge development into it and adjust it a bit.

Edru2 commented:
> Can we fix this? People don't have to be testers - they could want to downgrade to an earlier Mudlet version and losing their modules wouldn't be ideal.

:thinking: don't know of a clean way to fix this. I mean the issue is that the previous version (before this PR) aren't able to sync mpackage/zip modules and trash them. It won't trash xml modules.

vadi2 commented:
What if we add a separate flag for syncing modules in the save file? Then when you go back to an older version, the sync flag will not be enabled for an mpackage module.

Edru2 committed:
ensure backwards compatibility

this prevents trashing of zip modules if going back to a previous version (but the sync flag for zip modules will also be reset)

vadi2 commented on src/Host.cpp:
```diff
@@ -502,6 +503,34 @@ void Host::loadPackageInfo()
+    QFile::copy(filename, saveName + time);
+}
+
+void Host::writeModules(const QStringList &entry, const QString &moduleName, QString &filename)
```
If we're only using `entry[0]`, it would be better to pass the actual string we need - would be more understandable at a glance.

Edru2 replied:
Oh. I thought I did that already :man_facepalming:  will do

vadi2 commented on src/Host.cpp:
```diff
@@ -515,33 +544,25 @@ void Host::saveModules(int sync, bool backup)
     }
     modulesToWrite.clear();
-
+    //only reload the Module when the xml is ready
```
```suggestion
    // reload, or queue module reload for when xml is ready
```

vadi2 commented on src/Host.cpp:
```diff
@@ -502,6 +503,34 @@ void Host::loadPackageInfo()
+
+void Host::waitForHostXmlSave()
+{
+    auto myWriters = writers;
```
What's the need for this? The original code didn't have it

Edru2 replied:
This is needed so we wait for the host xml to save before we close the Host/Client.
It was in the original code as well I just took it out from waitForProfileSave and made it, its own function so I can use it somewhere else as well.

vadi2 replied:
I mean the `auto myWriters` bit?

Edru2 replied:
Oh that bit. 
I had crashes happening (sometimes) if I didn't do it this way.

The reason was that because of the async nature of this, in some cases writers (and futures) exactly got deleted right during the for loop was happening.

I tried some other ways to prevent that (like if writer/future == nullptr ...)
but the only way which prevented crashes was to put writers/futures into a variable right before the for loop was happening.

Not sure if in some cases this already caused crashes in Mudlet 4.13 (or earlier)

vadi2 replied:
OK makes sense, could you add a comment explaining it? Since off the shelf the code looks a bit odd

vadi2 commented on src/Host.h:
```diff
@@ -264,8 +264,7 @@ class Host : public QObject
     void startSpeedWalk(int sourceRoom, int targetRoom);
-    void saveModules(int sync, bool backup = true);
-    void reloadModule(const QString& reloadModuleName);
+    void reloadModule(const QString& reloadModuleName, const QString& syncHostName = QString());
```
`syncHostName` is the name of the profile we're syncing from, or syncing to?

Edru2 replied:
the name of the profile we're syncing from

vadi2 replied:
how about `syncingFromHost`?

vadi2 commented on src/Host.h:
```diff
@@ -711,6 +715,8 @@ private slots:
     // keeps track of all of the array writers we're currently operating with
     QHash<QString, XMLexport*> writers;
 
+    QFuture<void> moduleFuture;
```
```suggestion
    QFuture<void> mModuleFuture;
```

vadi2 commented on src/Host.cpp:
```diff
@@ -774,7 +807,14 @@ std::tuple<bool, QString, QString> Host::saveProfile(const QString& saveFolder,
-    saveModules(syncModules ? 1 : 0, saveName == QStringLiteral("autosave") ? false : true);
+    mWritingModules = true;
+    auto watcher = new QFutureWatcher<void>;
+    moduleFuture = QtConcurrent::run([=]() {
```
Since this waits not only for the modules, but the profile's own XML as well, the name should reflect that

vadi2 commented on src/Host.cpp:
```diff
@@ -502,6 +503,34 @@ void Host::loadPackageInfo()
+    auto myWriters = writers;
+    for (auto& writer : myWriters) {
+        auto myFutures = writer->saveFutures;
+        for (auto& future : myFutures) {
```
The name implies it only wants to wait for the profile XML to be saved, but it looks like it does modules as well?

Delwing commented on src/Host.cpp:
```diff
@@ -502,6 +503,34 @@ void Host::loadPackageInfo()
+    QFile::copy(filename, saveName + time);
+}
+
+void Host::writeModules(const QStringList &entry, const QString &moduleName, QString &filename)
```
```suggestion
void Host::writeModule(const QStringList &entry, const QString &moduleName, QString &filename)
```

Delwing commented on src/Host.h:
```diff
@@ -637,21 +637,25 @@ class Host : public QObject
     void createMapper(const bool);
     void removePackageInfo(const QString &packageName, const bool);
+    static void createModuleBackup(const QString &filename, const QString& saveName);
+    void writeModules(const QStringList &entry, const QString &moduleName, QString &filename);
```
```suggestion
    void writeModule(const QStringList &entry, const QString &moduleName, QString &filename);
```

Delwing commented on src/Host.cpp:
```diff
@@ -515,33 +544,25 @@ void Host::saveModules(int sync, bool backup)
-        writers.insert(filename_xml, writer);
-        writer->writeModuleXML(moduleName, filename_xml);
-
+        writeModules(entry, moduleName, filename);
```
```suggestion
        writeModule(entry, moduleName, filename);
```

vadi2 commented:
@demonnic could you retest since async saves got added? With some huge (10mb) packages as well, since it would not be unheard of for people to add soundpacks for example and such. I will do the same as well.

Delwing commented on src/XMLimport.cpp:
```diff
@@ -1600,8 +1600,14 @@ void XMLimport::readModulesDetailsMap(QMap<QString, QStringList>& map)
                 key = readElementText();
             } else if (name() == "filepath") {
```
Seems like reasonable suggestion.

vadi2 replied:
I'm not following it actually, can you explain? zipSync is not duplicated

Edru2 replied:
I don't understand it either. 
Does it mean to put name() == "filepath" and name() == zipSync into one line?

vadi2 commented on src/dlgModuleManager.cpp:
```diff
@@ -96,29 +96,17 @@ void dlgModuleManager::layoutModules()
             } else {
                 masterModule->setCheckState(Qt::Unchecked);
             }
             masterModule->setText(QString());
```
This text should be updated as well:

![image](https://user-images.githubusercontent.com/110988/140641240-36f8e939-39cc-4580-9800-e0ce1373c4ab.png)

vadi2 reviewed (commented):
Installed the Luminari UI http://luminarimud.com/download/LuminariGUI.mpackage and enabled sync, and while the mpackage with the xml was written, the config.lua and pictures were lost - is it the same for you?

Edru2 commented:
> Installed the Luminari UI http://luminarimud.com/download/LuminariGUI.mpackage and enabled sync, and while the mpackage with the xml was written, the config.lua and pictures were lost - is it the same for you?

What do you mean by they were lost?
I just tested it and in the mpackage file only the xml is updated.
Did you delete/move the mpackage file before saving?

vadi2 commented:
I did delete it, ok!

Edru2 committed:
apply suggestions

demonnic commented:
Retested and it works well for me =)

vadi2 reviewed (approved):
Thanks for your hard work on this 👍

vadi2 commented:
@demonnic can you re-approve?
